src/tyr/serializers.cc

```diff
@@ -179,6 +179,34 @@ waypoint(const valhalla::Location& location, bool is_tracepoint, bool is_optimiz
   return waypoint;
 }
 
+/*
+ * This function serializes the via_waypoints object on the leg.
+ */
+valhalla::baldr::json::MapPtr via_waypoint(const valhalla::Location& location,
+                                           const uint64_t geometry_idx) {
+
+  // Create a via waypoint to add to the array
+  auto via_waypoint = json::map({});
+  // Add distance in meters from the input location to the nearest
+  // point on the road used in the route
+  // TODO: since distance was normalized in thor - need to recalculate here
+  //       in the future we shall have store separately from score
+  via_waypoint
+      ->emplace("distance_from_leg_start",
+                json::fixed_t{to_ll(location.ll()).Distance(to_ll(location.path_edges(0).ll())), 3});
+
+  if (location.shape_index() == numeric_limits<uint32_t>::max()) {
+    // when waypoint is neither a break nor leg's starting/ending
+    // point (shape_index is uint32_t max), we assign null to its waypoint_index
+    via_waypoint->emplace("waypoint_index", static_cast<std::nullptr_t>(nullptr));
+  } else {
+    via_waypoint->emplace("waypoint_index", static_cast<uint64_t>(location.shape_index()));
```

Comment:
Is `shape_index` correct here? We want `waypoint_index` to be the index of the input location we're referring to?

Reply:
As we discussed in our meeting, this looks like it was mis-named in the proto so I have renamed to waypoint_index in the proto to fix any more confusion.